sender_pool.go

```diff
@@ -146,35 +171,49 @@ func (p *LineSenderPool) Acquire(ctx context.Context) (LineSender, error) {
 				return nil, errors.New("tcp/s not supported for pooled senders, use http/s only")
 			}
 		}
-		return newHttpLineSender(conf)
+		s, err = newHttpLineSender(conf)
+	}
+
+	if err != nil {
+		return nil, err
 	}
 
+	p.numSenders++
+
+	ps := &pooledSender{
+		pool:    p,
+		wrapped: s,
+	}
+	atomic.AddUint64(&ps.tick, 1)
+	return ps, nil
 }
 
-// Release flushes the LineSender and returns it back to the pool. If the pool
-// is full, the sender is closed and discarded. In cases where the sender's
-// flush fails, it is not added back to the pool.
-func (p *LineSenderPool) Release(ctx context.Context, s LineSender) error {
-	// If there is an error on flush, do not add the sender back to the pool
-	if err := s.Flush(ctx); err != nil {
-		return err
+func (p *LineSenderPool) free(ctx context.Context, ps *pooledSender) error {
+	var flushErr error
+
+	if ps.dirty {
+		flushErr = ps.Flush(ctx)
 	}
 
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	for i := range p.senders {
-		if p.senders[i] == s {
-			return fmt.Errorf("LineSender %p has already been released back to the pool", s)
-		}
+	if flushErr != nil {
+		// Failed to flush, close and call it a day
+		p.numSenders--
+		closeErr := ps.wrapped.Close(ctx)
+		return fmt.Errorf("%s %w", flushErr, closeErr)
 	}
 
-	if p.closed || len(p.senders) >= p.maxSenders {
-		return s.Close(ctx)
+	if ps.dirty || p.closed {
+		// Previous error or closed pool, close and call it a day
+		p.numSenders--
+		return ps.wrapped.Close(ctx)
 	}
 
-	p.senders = append(p.senders, s)
-
+	p.freeSenders = append(p.freeSenders, ps)
+	// Notify free sender waiters, if any
+	p.cond.Broadcast()
```

Comment:
If you're going to short-circuit `return`s earlier in this method, I think you should `defer` this `p.cond.Broadcast()` call to ensure that all goroutines with blocked senders get properly cleaned up. Even if the pool is closed, we don't want to leak goroutines

Reply:
Good spot. Fixed in 1b442eb91b186904cb7925a9340484e8a245d357